src/openssl_tls.rs

```diff
@@ -0,0 +1,212 @@
+#![cfg(feature = "tls-openssl")]
+
+use std::{task::Poll};
+
+use http::{Request, Uri};
+use hyper::client::HttpConnector;
+use hyper_openssl::HttpsConnector;
+use openssl::{
+    error::ErrorStack,
+    pkey::PKey,
+    ssl::{SslConnector, SslMethod},
+    x509::X509,
+};
+use tokio::sync::mpsc::{Receiver, Sender};
+use tokio_stream::wrappers::ReceiverStream;
+use tonic::{
+    body::BoxBody,
+    transport::{Channel, Endpoint},
+};
+use tower::{balance::p2c::Balance, buffer::Buffer, discover::Change, load::Load, Service};
+
+use super::error::Result;
+
+pub type SslConnectorBuilder = openssl::ssl::SslConnectorBuilder;
+pub type OpenSslResult<T> = std::result::Result<T, ErrorStack>;
+// Below are some type alias for make clearer types.
+pub type TonicRequest = Request<BoxBody>;
+pub type Buffered<T> = Buffer<T, TonicRequest>;
+pub type Balanced<T> = Balance<T, TonicRequest>;
+pub type OpenSslChannel = Buffered<Balanced<OpenSslDiscover<Uri>>>;
+/// OpenSslDiscover is the backend for balanced channel based on OpenSSL transports.
+/// Because `Channel::balance` doesn't allow us to provide custom connector, we must implement ourselves' balancer...
+pub type OpenSslDiscover<K> = ReceiverStream<Result<Change<K, FairLoaded<Channel>>>>;
+
+#[derive(Clone)]
+pub struct OpenSslConnector(HttpsConnector<HttpConnector>);
+
+impl std::fmt::Debug for OpenSslConnector {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("OpenSslConnector").finish()
+    }
+}
+
+impl OpenSslConnector {
+    pub fn create_default() -> OpenSslResult<Self> {
+        let conf = OpenSslClientConfig::default();
+        conf.build()
+    }
+}
+
+#[cfg(feature = "tls")]
+compile_error!(concat!(
+    "**You should only enable one of `tls` and `tls-openssl`.** Reason: ",
+    "For now, `tls-openssl` would take over the transport layer (sockets) to implement TLS based connection. ",
+    "As a result, once using with `tonic`'s internal TLS implementation (which based on `rustls`), ", 
+    "we may create TLS tunnels over TLS tunnels or directly fail because of some sorts of misconfiguration.")
+);
+
+/// `FairLoaded` is a simple wrapper over channels that provides nothing about work load.
+/// (Which would lead to a complete "fair" scheduling?)
+#[repr(transparent)]
+pub struct FairLoaded<S>(S);
+
+impl<S> Load for FairLoaded<S> {
+    type Metric = usize;
+
+    fn load(&self) -> Self::Metric {
+        0
+    }
+}
```

Comment:
This is simply [`tower::load::Constant<S>`](https://docs.rs/tower/latest/tower/load/struct.Constant.html) with `Constant::new(inner, 0)`